### ISSUE TITLE
random SIGSEGV when resolving queries

### ISSUE THREAD

**Issue by iduartgomez** [iduartgomez/simag#1]
lldb backtrack sample:

```
* thread #2: tid = 24319, 0x0000555555652a1e simag-da6d0995a3679276`_$LT$core..hash..sip..Hasher$LT$S$GT$$u20$as$u20$core..hash..Hasher$GT$::write::hcf9671714aef2b41 + 302 at sip.rs:100, name = 'agent::kb::test', stop reason = signal SIGSEGV: invalid address (fault address: 0x0)
  * frame #0: 0x0000555555652a1e simag-da6d0995a3679276`_$LT$core..hash..sip..Hasher$LT$S$GT$$u20$as$u20$core..hash..Hasher$GT$::write::hcf9671714aef2b41 + 302 at sip.rs:100
    frame #1: 0x000055555564b41d simag-da6d0995a3679276`_$LT$core..hash..sip..SipHasher13$u20$as$u20$core..hash..Hasher$GT$::write::h1bd19b4cd53305e7 + 61 at sip.rs:247
    frame #2: 0x00005555556bc5ed simag-da6d0995a3679276`_$LT$std..collections..hash..map..DefaultHasher$u20$as$u20$core..hash..Hasher$GT$::write::h11bb3e1110f1e921 + 61 at map.rs:2043
    frame #3: 0x00005555556079fd simag-da6d0995a3679276`core::hash::impls::_$LT$impl$u20$core..hash..Hash$u20$for$u20$str$GT$::hash::h7939a52de837a50c + 93 at mod.rs:342
    frame #4: 0x0000555555724b64 simag-da6d0995a3679276`_$LT$simag..agent..kb..VarAssignment$LT$$u27$a$GT$$u20$as$u20$core..hash..Hash$GT$::hash::h677dfd56dbbd8561 + 52 at kb.rs:1640
    frame #5: 0x0000555555607cf0 simag-da6d0995a3679276`core::hash::impls::_$LT$impl$u20$core..hash..Hash$u20$for$u20$$RF$$u27$a$u20$T$GT$::hash::h8065a7ac0369e324 + 48 at mod.rs:393
    frame #6: 0x0000555555607d4e simag-da6d0995a3679276`core::hash::impls::_$LT$impl$u20$core..hash..Hash$u20$for$u20$$LP$A$C$$u20$B$RP$$GT$::hash::h12fc8b007a2ed7d2 + 78 at mod.rs:369
    frame #7: 0x00005555556078e7 simag-da6d0995a3679276`core::hash::Hash::hash_slice::h1bd1315c1acfab32 + 151 at mod.rs:142
    frame #8: 0x0000555555607b6b simag-da6d0995a3679276`core::hash::impls::_$LT$impl$u20$core..hash..Hash$u20$for$u20$$u5b$T$u5d$$GT$::hash::ha42f723341448ccb + 91 at mod.rs:385
    frame #9: 0x000055555564a342 simag-da6d0995a3679276`_$LT$collections..vec..Vec$LT$T$GT$$u20$as$u20$core..hash..Hash$GT$::hash::h628c375099f9e5e3 + 66 at vec.rs:1296
    frame #10: 0x00005555556348da simag-da6d0995a3679276`_$LT$alloc..arc..Arc$LT$T$GT$$u20$as$u20$core..hash..Hash$GT$::hash::h8d63902de903a57c + 58 at arc.rs:963
    frame #11: 0x00005555555d5444 simag-da6d0995a3679276`std::collections::hash::table::make_hash::h1117b5017a6a25de + 100 at table.rs:159
    frame #12: 0x0000555555678fdd simag-da6d0995a3679276`_$LT$std..collections..hash..map..HashMap$LT$K$C$$u20$V$C$$u20$S$GT$$GT$::make_hash::h1ffbead64d99e325 + 45 at map.rs:490
    frame #13: 0x0000555555676a48 simag-da6d0995a3679276`_$LT$std..collections..hash..map..HashMap$LT$K$C$$u20$V$C$$u20$S$GT$$GT$::search::hab8c92ceadf04d9c + 56 at map.rs:500
    frame #14: 0x0000555555655117 simag-da6d0995a3679276`_$LT$std..collections..hash..map..HashMap$LT$K$C$$u20$V$C$$u20$S$GT$$GT$::contains_key::h1f4bc828b63b58fc + 55 at map.rs:1096
    frame #15: 0x00005555556370bd simag-da6d0995a3679276`_$LT$std..collections..hash..set..HashSet$LT$T$C$$u20$S$GT$$GT$::contains::h322853f7775501cd + 45 at set.rs:449
    frame #16: 0x00005555557241e3 simag-da6d0995a3679276`_$LT$simag..agent..kb..ArgsProduct$LT$$u27$a$GT$$u20$as$u20$core..iter..iterator..Iterator$GT$::next::h39a5a337986654fc + 819 at kb.rs:1538
    frame #17: 0x000055555571db94 simag-da6d0995a3679276`simag::agent::kb::InfTrial::unify::h42733634a1e86183 + 3940 at kb.rs:1178
    frame #18: 0x000055555571e8f0 simag-da6d0995a3679276`simag::agent::kb::InfTrial::unify::h42733634a1e86183 + 7360 at kb.rs:1227
    frame #19: 0x000055555571e8f0 simag-da6d0995a3679276`simag::agent::kb::InfTrial::unify::h42733634a1e86183 + 7360 at kb.rs:1227
```

this is probably due to a short lived reference gotten through unsafe deref

**Comment by iduartgomez** [comment]
An other one happening elsewhere. Both involve VarAssignment structs, which have internal references which could be outlived (double check for unsafe code)?
```
* thread #2: tid = 28349, 0x0000555555682d8b simag-da6d0995a3679276`_$LT$std..collections..hash..table..Bucket$LT$K$C$$u20$V$C$$u20$M$GT$$GT$::peek::h4efdd49454125d42 + 75 at table.rs:291, name = 'agent::kb::test', stop reason = signal SIGSEGV: invalid address (fault address: 0x40)
    frame #0: 0x0000555555682d8b simag-da6d0995a3679276`_$LT$std..collections..hash..table..Bucket$LT$K$C$$u20$V$C$$u20$M$GT$$GT$::peek::h4efdd49454125d42 + 75 at table.rs:291
(lldb) bt
* thread #2: tid = 28349, 0x0000555555682d8b simag-da6d0995a3679276`_$LT$std..collections..hash..table..Bucket$LT$K$C$$u20$V$C$$u20$M$GT$$GT$::peek::h4efdd49454125d42 + 75 at table.rs:291, name = 'agent::kb::test', stop reason = signal SIGSEGV: invalid address (fault address: 0x40)
  * frame #0: 0x0000555555682d8b simag-da6d0995a3679276`_$LT$std..collections..hash..table..Bucket$LT$K$C$$u20$V$C$$u20$M$GT$$GT$::peek::h4efdd49454125d42 + 75 at table.rs:291
    frame #1: 0x00005555555c74ac simag-da6d0995a3679276`std::collections::hash::map::search_hashed::h7a37b7b19701251c + 444 at map.rs:371
    frame #2: 0x000055555567657a simag-da6d0995a3679276`_$LT$std..collections..hash..map..HashMap$LT$K$C$$u20$V$C$$u20$S$GT$$GT$::search::hf91161ddb1eff522 + 154 at map.rs:501
    frame #3: 0x0000555555662a07 simag-da6d0995a3679276`_$LT$std..collections..hash..map..HashMap$LT$K$C$$u20$V$C$$u20$S$GT$$GT$::get::h48349ec692dd0515 + 71 at map.rs:1070
    frame #4: 0x0000555555723773 simag-da6d0995a3679276`simag::agent::kb::VarAssignment::get_relationship::h220b3ab40c8dbc46 + 99 at kb.rs:1603
    frame #5: 0x0000555555708bd3 simag-da6d0995a3679276`simag::lang::common::FuncDecl::equal_to_grounded::h4284ae69f4cca262 + 1043 at common.rs:985
    frame #6: 0x00005555557042eb simag-da6d0995a3679276`simag::lang::common::Assert::equal_to_grounded::h6b5fde371c1a9d23 + 107 at common.rs:628
    frame #7: 0x00005555556dae27 simag-da6d0995a3679276`simag::lang::logsent::LogicAtom::solve::h4c24b6f7f5a34fcd + 135 at logsent.rs:683
    frame #8: 0x00005555556db4bb simag-da6d0995a3679276`simag::lang::logsent::Particle::solve::he11255bf764344ac + 891 at logsent.rs:740
    frame #9: 0x00005555556d9419 simag-da6d0995a3679276`simag::lang::logsent::LogicConjunction::solve::he1e5c8c3fb89a8b0 + 425 at logsent.rs:529
    frame #10: 0x00005555556db29e simag-da6d0995a3679276`simag::lang::logsent::Particle::solve::he11255bf764344ac + 350 at logsent.rs:735
    frame #11: 0x00005555556d9419 simag-da6d0995a3679276`simag::lang::logsent::LogicConjunction::solve::he1e5c8c3fb89a8b0 + 425 at logsent.rs:529
    frame #12: 0x00005555556db29e simag-da6d0995a3679276`simag::lang::logsent::Particle::solve::he11255bf764344ac + 350 at logsent.rs:735
    frame #13: 0x00005555556d9419 simag-da6d0995a3679276`simag::lang::logsent::LogicConjunction::solve::he1e5c8c3fb89a8b0 + 425 at logsent.rs:529
    frame #14: 0x00005555556db29e simag-da6d0995a3679276`simag::lang::logsent::Particle::solve::he11255bf764344ac + 350 at logsent.rs:735
    frame #15: 0x00005555556d7089 simag-da6d0995a3679276`simag::lang::logsent::LogicIndCond::solve::h0e0f58986a1c341b + 169 at logsent.rs:309
    frame #16: 0x00005555556db450 simag-da6d0995a3679276`simag::lang::logsent::Particle::solve::he11255bf764344ac + 784 at logsent.rs:739
    frame #17: 0x00005555556d416e simag-da6d0995a3679276`simag::lang::logsent::LogSentence::solve::h56d1ed51ff7eed5b + 446 at logsent.rs:110
    frame #18: 0x000055555571d048 simag-da6d0995a3679276`simag::agent::kb::InfTrial::unify::ha723b04b8dd021bd + 5336 at kb.rs:1193

```

**Comment by iduartgomez** [comment]
fixed in # d4f79a65a4d6f38b978788cac2398093d29a994d
